extension_distribution.md

```diff
@@ -0,0 +1,296 @@
+
+You install them into VSCode itself (rather than a particular repo).
+
+They typically add things to the Command Palette. But there are also extensions
+that manage other tools on your machine (e.g., the Minikube extension makes sure
+Minikube is up to date).
+
+#### Browser extensions
+
+These typically come in two flavors:
+
+- Extensions that add buttons or capapilities to the browser itself (like [tabby
+  cat](https://chrome.google.com/webstore/detail/tabby-cat/mefhakmgclhhfbdadeojlkbllmecialg?hl=en),
+  the best extension)
+
+- Extensions that modify pages you visit, like Google Translate.
+
+#### Gradle Plugins (https://docs.gradle.org/current/userguide/implementing_gradle_plugins.html)
+
+Add custom task types to your build system, or decorate existing tasks with new
+outputs. Here are some examples: https://github.com/ksoichiro/awesome-gradle
+
+#### Bazel Extensions (https://docs.bazel.build/versions/main/skylark/concepts.html)
+
+Mostly focused on adding macros and custom build rules.
+
+#### Helm Chart Repositories (https://helm.sh/docs/helm/helm_repo/)
+
+The Helm repo command lets you install new repositories of charts. Each chart is a set of resources
+to deploy. These charts can add new types and operators to your Kubernetes cluster.
+
+The chart repos are globally installed, but the charts are installed to a particular cluster.
+
+#### yarn add / yarn global add
+
+In the javascript ecoysystem, packages are installed locally (in the current
+repo) or globally (in a shared global package tracker). Packages can contain
+both library code that you have to explicitly import, but also contains
+executables that you can run.
+
+## Proposal
+
+We'll add three new data types to the APIServer:
+
+- `ExtensionRepo`: A repository of Tilt extensions.
+- `Extension`: A globally installed extension.
+- `Tiltfile`: A Tiltfile invocation.
+
+Most data types in the Tilt APIServer are stored in-memory only, and disappear
+when Tilt is killed. The `ExtensionRepo` and `Extension` types are
+different. They're persisted on-disk across Tilt sessions under ~/.tilt-dev.
+
+### ExtensionRepo
+
+```go
+type ExtensionRepoSpec struct {
+  // The URL of the repo.
+  // 
+  // Allowed:
+  // https: URLs that point to a public git repo
+  // file: URLs that point to a location on disk.
+  URL string
+}
+```
+
+```go
+type ExtensionRepoStatus struct {
+  // Contains information about any problems loading the repo.
+  Error string
+
+  // The last time the repo was fetched and checked for validity.
+  LastFetched metav1.Time
+}
+```
+
+### Extension
+
+The name of an extension MUST have the form `{ExtensionRepo.Name}/{path}` where
+the first part of the name points to an existing `ExtensionRepo`.
+
+If an `ExtensionRepo` is deleted, all the `Extension` objects with the same prefix
+are also deleted.
+
+Each `Extension` will manage a child object, a `Tiltfile`, which manages
+the extension execution for a particular Tilt session.
```

Comment:
Making sure I understand the model here, the `ExtensionReconciler` will create a `TiltfileSpec` for `extension.spec.path + "/Tiltfile"`?

And that Tiltfile will execute in its own context and is generally expected to e.g. fork off a Bash script that runs the reactive loop?

Reply:
I had the same question; maybe it'd be worth considering renaming Tiltfile in this context to something different? 🤔

Reply:
(Ah ok I followed L's self-advice and read the example and this makes more sense now)

Reply:
:thinking: 

ya, Tiltfile isn't a great name. what about TiltfileJob? TiltfileCmd? TiltfileConfig?

Reply:
FWIW I'm not wholly opposed to the name, I was more trying to understand execution flow - at first I didn't realize the `Tiltfile` object was the extension execution context; I thought it was _your_ app `Tiltfile`. (I had assumed the extension would just have some sort `entrypoint` like a Dockerfile that the reconciler would turn directly into `Cmd`s.)

More names to throw into the mix: `TiltfileExec` / `TiltfileRun` / `TiltfileLoad`. I have no real preference

Reply:
ahhhhhh